CONTRIBUTING.md

```diff
@@ -0,0 +1,584 @@
+
+### Converting CUDA code to HIP code
+
+This script calls the `hipify` script provided by HIP converting the CUDA syntax
+to HIP syntax. Additionally, it also automatically replaces the instances of
+CUDA with HIP as appropriate. Hence, this script can be called on a Ginkgo CUDA
+file.
+
+
+## Writing Tests
+
+Ginkgo uses the [GTest framework](https://github.com/google/googletest) for the
+unit test framework within Ginkgo. Writing good tests are extremely important to
+verify the functionality of the new code and to make sure that none of the
+existing code has been broken.
+
+### Testing know-how
+
+* GTest provides a [comprehensive documentation](https://github.com/google/googletest/blob/master/googletest/docs/primer.md) of the functionality available within Gtest.  
+* Reduce code duplication with [Testing Fixtures,
+  `TEST_F`](https://github.com/google/googletest/blob/master/googletest/docs/primer.md#test-fixtures-using-the-same-data-configuration-for-multiple-tests-same-data-multiple-tests)
+* Write templated tests using
+  [`TYPED_TEST`](https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#typed-tests).
+
+### Some general rules.
+
+* Unit tests must follow the [KISS
+  principle](https://en.wikipedia.org/wiki/KISS_principle).
+* Unit tests must follow the [AAA](http://wiki.c2.com/?ArrangeActAssert)
+  pattern, and a single blank line must appear between consecutive "A" sections.
+
+### Writing tests for kernels
+
+* Reference kernels, kernels on the `ReferenceExecutor`, are meant to be single
+  threaded reference implementations. Therefore, tests for reference kernels
+  need to be performed with data that can be as small as possible. For example,
+  matrices lesser than 5x5 are acceptable. This allows the reviewers to verify
+  the results for exactness with tools such as MATLAB.
+* OpenMP, CUDA and HIP kernels have to be tested against the reference kernels.
+  Hence data for the tests of these kernels can be generated in the test files
+  using helper functions or by using external files to be read through the
+  standard input. In particular for CUDA and HIP, the data size should be at
+  least bigger than the architecture's warp size to ensure there is no corner
+  case in the kernels.
+
+
+## Documentation style
+
+Documentation uses standard Doxygen.
+
+###  Developer targeted notes
+Make use of `@internal` doxygen tag. This can be used for any comment which is
```

Comment:
nit: missing empty line
```suggestion
###  Developer targeted notes

Make use of `@internal` doxygen tag. This can be used for any comment which is
```